Neos.Flow/Classes/Mvc/Routing/Router.php

```diff
@@ -267,9 +276,112 @@ protected function createRoutesFromConfiguration()
             }
             $this->routes[] = $route;
         }
+
         $this->routesCreated = true;
     }
 
+    /**
+     * Emits a signal when the router configuration has been loaded
+     *
+     * This signal can be used to add routes during runtime. In the slot make sure to receive the
+     * $routerConfiguration array by reference so you can alter it.
+     *
+     * @param array $routerConfiguration The router configuration
+     * @return void
+     * @Flow\Signal
+     */
+    protected function emitConfigurationLoaded(array &$routerConfiguration): void
+    {
+    }
+
+    public function configureRouteAnnotatedMethods(array &$routerConfiguration): void
+    {
+        $annotatedRoutes = static::resolveRouteAnnotatedMethods($this->objectManager);
+        $routerConfiguration = array_merge($routerConfiguration, array_values($annotatedRoutes));
```

Comment:
I need to look into this more thoroughly (hopefully later today or tomorrow). But IMO the `Routes.yaml` should have precedence so that the routing can be changed for 3rd party packages that define a default behavior via annotations.

And, related, it should be possible to override single routes (instead of adding more specific ones) I think (we could use the route name for that). Just a rough idea, but IMO this is a critical point in order to avoid non-canonical URL behavior (i.e. two URLs maching the same route)

Reply:
> And, related, it should be possible to override single routes (instead of adding more specific ones) I think (we could use the route name for that).

Sound legit. It will require us to make a "name" or "identifier" of a route required (or build from a hash of the uriPattern?)

> Just a rough idea, but IMO this is a critical point in order to avoid non-canonical URL behavior (i.e. two URLs matching the same route)

Router::createRoutesFromConfiguration makes sure the same uriPattern is not registered twice

Reply:
> 3rd party packages that define a default behavior via annotations

IMO thats not the use-case for annotation routes and I would not recommend doing that. Defaults should be given in global settings and exceptions be defined inplace. A developer looking at a Controller with a `@Flow\Route("my/path")` annotation should always expect this method to be reachable exactly under that route.

> > Just a rough idea, but IMO this is a critical point in order to avoid non-canonical URL behavior (i.e. two URLs matching the same route)
>
> Router::createRoutesFromConfiguration makes sure the same uriPattern is not registered twice

Non-canonical means that two distinct uriPatterns would match a single action, which would lead to duplicate content (SEO issue). I think we could also check that to some degree, though we probably can't completely prevent that. It's already possible right now, so maybe adding such a check should be an own PR.

Reply:
you're right. It's probably not necessary to override single routes specifically. But I still think that it should be possible to add more specific routes via YAML (read: routes that are evaluated first)

Reply:
Under which circumstances would you consider a YAML route more specific than one annotated directly on the method? The only case I could think of would be:
- settings route specifies URI constraints/http method and a single controller action (no `{@controller}` and `{@action}` parameter)
- the route on the method specifies no constraint/http method
- the incoming request exactly matches the constraint/http method

So with that we would need to assign specifity levels to package/controller/action/http method/uri constraint, then merge based on that level? Sounds like a lot of complexity, which I would rather add in a second step. In most cases, a Route annotation has higher specifity.